jax/_src/numpy/lax_numpy.py

```diff
@@ -3304,7 +3308,8 @@ def array(object: Any, dtype: DTypeLike | None = None, copy: bool = True,
     out = np.array(object) if copy else np.asarray(object)
   else:
     raise TypeError(f"Unexpected input type for array: {type(object)}")
-
+  if device is not None:
+    out = jax.device_put(out, device=device)
```

Comment:
This will convert `out` from whatever it is to a JAX array, potentially with the wrong type. To avoid that intermediate allocation, I think we should call `device_put` on `out_array` below.

Reply:
maybe we don't need to do this at all? We should not incur the cost of an extra transfer. We should instead pass the device down to _convert_element_type and then to jit after.

Reply:
Hey @vfdev-5 Yash and I just chatted a bit offline, and I think his plan is good here: basically, we need to focus first on `array` and `asarray`, pushing the device logic deep (probably into `_convert_element_type`). Then in the other APIs touched here and in related PRs, we can use the `array` logic rather than re-implementing the same logic in each function. What do you think?

Reply:
https://github.com/google/jax/pull/22331 is my PR which adds `sharding` to `convert_element_type_p` primitive.